python/ray/experimental/channel/common.py

```diff
@@ -380,26 +391,36 @@ def write(self, val: Any, timeout: Optional[float] = None) -> None:
 
     def close(self) -> None:
         self._closed = True
-        self._output_channel.close()
+        for channel in self._output_channels:
+            channel.close()
 
 
 @DeveloperAPI
 class SynchronousWriter(WriterInterface):
     def start(self):
-        self._output_channel.ensure_registered_as_writer()
-        pass
+        for channel in self._output_channels:
+            channel.ensure_registered_as_writer()
 
     def write(self, val: Any, timeout: Optional[float] = None) -> None:
-        self._output_channel.write(val, timeout)
+        for i, channel in enumerate(self._output_channels):
+            idx = self._output_idxs[i]
```

Comment:
Nit: Instead of storing `self._output_idxs`, it seems a bit cleaner to just use `None` in `self._output_channels` to indicate that a return value is not used. Then, you can simply use `zip(val, self._output_channels)` instead of looking up each return value individually. So it can look like this:

```python
self._output_channels: List[Optional[ChannelInterface]]
```

instead of:
```python
self._output_channels: List[ChannelInterface]
self._output_idxs: List[Optional[int]]
```

Reply:
If `output_idx` is None, it means the complete `val` is returned. If `output_idx` is not None, it means `val[output_idx]` is returned. `None` is used to decide a complete or a slice value is returned.

Reply:
Hmm I don't think it is too complicated to require `self._output_channels` be ordered. Actually it seems simpler to me because then we can remove the extra layer of indirection (`self._output_idxs`). It also pushes more of the logic to compile-time instead of run-time: basically you could take almost the same code here and move it into the `WriterInterface` constructor.

I would still push for this as a simplification but can do it in a follow-up PR in the interest of getting this merged.